데이터 출처: 공공데이터 포털<br>
https://www.data.go.kr/data/15034107/openapi.do?recommendDataYn=Y

# 공무원 맞춤형복지 카드사용현황

|항목명(국문)|항목명(영문)|항목크기|항목구분|샘플데이터|항목설명|
|---|---|---|---|---|---|
|인증키|serviceKey|100|필수|인증키(URL Encode)|공공데이터포털에서 받은 인증키|
|응답 데이터 형식|ViewType|4|옵션|xml 또는 json|응답 데이터 형식 생략 시 기본값은 xml|
|한 페이지 결과 수|numOfRows|3|필수|10|한 페이지 결과 수 최대값은 100|
|페이지 번호|pageNo|4|필수|1|페이지 번호 0보다 큰 값|
|사용연월|useYymm|6|필수|201801|검색을 원하는 사용연월|
|성별|sxClNm|2|필수|여(URL Encode)|검색을 원하는 성별(남 또는 여)|
|연령별|agePrd|4|옵션|30대(URL Encode)|검색을 원하는 연령대|
|1차분류코드|fstCtgy|4|옵션|D|검색을 원하는 1차분류코드 e)코드 명세 참조|
|1차분류코드명|fstCtgyNm|100|옵션|여가생활(URL Encode)|검색을 원하는 1차분류코드명 e)코드 명세 참조|
|2차분류코드|sndCtgy|4|옵션|D13|검색을 원하는 2차분류코드 e)코드 명세 참조|
|2차분류코드명|sndCtgyNm|100|옵션|전자상거래(URL Encode)|검색을 원하는 2차분류코드명 e) 코드 명세 참조|

|항목명(영문)|항목명(국문)|항목크기|항목구분|샘플데이터|항목설명|
|---|---|---|---|---|---|
|resultCode|응답코드|2|1|0|응답코드이며, "2. Open API 응답코드정리" 참조|
|resultMsg|응답메시지|50|1|NORMAL_SERVICE|응답메시지이며, "2. Open API 응답코드정리" 참조|
|numOfRows|한 페이지 결과 수|3|1|10|한 페이지당 표출 데이터 수|
|pageNo|페이지 번호|4|1|1|페이지 번호|
|totalCount|데이터 총 개수|4|1|1500|데이터 총 개수|
|no|일련번호|8|1|1|일련번호|
|useYymm|사용연월|6|1|201801|사용연월|
|sxClNm|성별|2|1|남|성별 (남 또는 여)|
|agePrd|연령별|4|1|40대|연령별(10대, 20대 등)|
|fstCtgy|1차분류코드|4|1|A|1차분류코드 e) 코드 명세 참조|
|fstCtgyNm|1차분류코드명|100|1|가정친화|1차분류코드명 e) 코드 명세 참조|
|sndCtgy|2차분류코드|4|1|A01|2차분류코드 e) 코드 명세 참조|
|sndCtgyNm|2차분류코드명|100|1|가족식사|2차분류코드명 e) 코드 명세 참조|
|crdCoBzkNm|업종명칭|100|1|간이음식점|업종명칭|
|crdUseCnt|사용건수|8|1|5|사용건수|
|crdUseAmt|사용금액|13|1|120500|사용금액|

|1차분류코드|속성|
|---|---|
|A|가정친화|
|B|건강관리|
|C	|자기계발|
|D	|여가생활|
|E	|온누리상품권|
|G	|전통(재래)시장|

|2차분류코드|속성|2차분류코드|속성|
|---|---|---|---|
|A01|가족식사|C01|학원수강|
|A02|음료/식품|C02|학습자재/도서구매|
|A03|가전제품|D01|레저/스포츠|
|A04|컴퓨터/통신|D02|미용|
|A05|생활용품|D04|여행/교통|
|A08|가구/인테리어/건축자재|D05|숙박시설|
|A09|의류/잡화|D06|여가활용|
|A10|자녀보육|D07|취미생활|
|A11|차량서비스/차량연료|D08|오락/문화/관람|
|A13|결혼/장례서비스지원|D09|기념품|
|A16|기부금|D10|공예/토산품|D10|공예/토산품|
|A17|제조/도매업관리|D11|관광(외국인전용점)|
|A18|선물용품|D12|생활편의시설(유통)|
|A19|유아용품|D13|전자상거래|
|B01|병의원진료|D14|운송/임대|
|B02|약국및의료기기구매|D15|용역서비스|
|B03|건강식품|E01|온누리상품권(개인구매)|
|B04|보험/금융|E03|온누리상품권(개인의무구매)|
|G01|재래시장상품권|G02|전통(재래)시장 사용|

# 데이터 전처리 - NLP 포함

> ### NLP 엔진은 Mecab 사용  
  외부 설치 작업 필요 (Konlpy만으로는 완전 설치 불가)

> 연월을 쪼개서 새로운 컬럼 생성 필요

### 패키지 로드 + NLP 엔진 작동 테스트

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib as mpl

In [125]:
from tensorflow import keras
from keras.preprocessing.text import Tokenizer
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
import re

In [2]:
# 기본 한글 폰트 셋팅
mpl.rcParams['font.family']='Malgun Gothic'

In [128]:
# Mecab 로드
from konlpy.tag import Mecab
mecab = Mecab(dicpath=r"C:\mecab\mecab-ko-dic")

In [129]:
# Mecab 작동 테스트
print('형태소 단위: ',mecab.morphs("안녕하세요 여러분"))
print('명사 단위: ',mecab.nouns("안녕하세요 여러분"))
print('품사 태깅: ',mecab.pos("안녕하세요 여러분"))

형태소 단위:  ['안녕', '하', '세요', '여러분']
명사 단위:  ['안녕', '여러분']
품사 태깅:  [('안녕', 'NNG'), ('하', 'XSV'), ('세요', 'EP+EF'), ('여러분', 'NP')]


---

### 데이터 로드

In [4]:
# CSV 파일 로드
raw_df = pd.read_csv('Data/카드이용내역.csv')
df=raw_df.copy()
df

crdCoBzkNm   sndCtgyNm sndCtgy sxClNm  useYymm fstCtgyNm agePrd  no  \
0               양식        가족식사     A01      남   201801      가정친화    10대   1   
1         일반음식점 기타        가족식사     A01      남   201801      가정친화    10대   2   
2        일식  /생선회집        가족식사     A01      남   201801      가정친화    10대   3   
3               중식        가족식사     A01      남   201801      가정친화    10대   4   
4       커피  /음료전문점        가족식사     A01      남   201801      가정친화    10대   5   
...            ...         ...     ...    ...      ...       ...    ...  ..   
296897  일반.치과.한방병원       병의원진료     B01      여   201912      건강관리    80대   4   
296898        종합병원       병의원진료     B01      여   201912      건강관리    80대   5   
296899          약국   약국및의료기기구매     B02      여   201912      건강관리    80대   6   
296900        화장품점          미용     D02      여   201912      여가생활    80대   7   
296901        슈퍼마켓  생활편의시설(유통)     D12      여   201912      여가생활    80대   8   

        crdUseAmt  crdUseCnt fstCtgy  
0           12000          1       A  
1          100000          4       A  
2           12000          1       A  
3           30000          2       A  
4           30900          4       A  
...           ...        ...     ...  
296897      49700          1       B  
296898      95100          1       B  
296899     122300          1       B  
296900      17000          1       D  
296901     134360          8       D  

[296902 rows x 11 columns]

In [7]:
df[df['crdCoBzkNm']=='안경,콘텍트렌즈']

crdCoBzkNm  sndCtgyNm sndCtgy sxClNm  useYymm fstCtgyNm agePrd   no  \
1944     안경,콘텍트렌즈       가전제품     A03      남   201801      가정친화    20대  392   
2208     안경,콘텍트렌즈  약국및의료기기구매     B02      남   201801      건강관리    20대  656   
2334     안경,콘텍트렌즈       가전제품     A03      남   201802      가정친화    20대   75   
2583     안경,콘텍트렌즈  약국및의료기기구매     B02      남   201802      건강관리    20대  324   
3118     안경,콘텍트렌즈       가전제품     A03      남   201803      가정친화    20대  184   
...           ...        ...     ...    ...      ...       ...    ...  ...   
294427   안경,콘텍트렌즈       가전제품     A03      여   201904      가정친화    70대   46   
294515   안경,콘텍트렌즈  약국및의료기기구매     B02      여   201905      건강관리    70대   51   
294979   안경,콘텍트렌즈  약국및의료기기구매     B02      여   201909      건강관리    70대   63   
295920   안경,콘텍트렌즈  약국및의료기기구매     B02      여   202005      건강관리    70대   64   
296397   안경,콘텍트렌즈  약국및의료기기구매     B02      여   202009      건강관리    70대   61   

        crdUseAmt  crdUseCnt fstCtgy  
1944      1712000         20       A  
2208     14126000        148       B  
2334      1377000         17       A  
2583     10163200        110       B  
3118      1840500         19       A  
...           ...        ...     ...  
294427      50000          1       A  
294515     530000          4       B  
294979      40000          1       B  
295920     150000          1       B  
296397      42000          1       B  

[761 rows x 11 columns]

---

### 커스텀 함수

In [131]:
# 데이터프레임을 텍스트 파일로 저장하는 함수
def DFtofile(DFtofile,resname):
    a=pd.DataFrame(DFtofile)
    a.to_csv('_res/'+resname+'.txt')

In [132]:
# 피벗테이블을 통해 중복 Y값이 있는 카테고리를 탐색
def pivot_check(df,names):
    temp=df[['crdCoBzkNm','sndCtgyNm']].groupby(['crdCoBzkNm', 'sndCtgyNm'])[['sndCtgyNm']].count()
    temp.columns=['count']
    temp.reset_index(inplace=True)
    #temp.pivot(index='crdCoBzkNm', columns='sndCtgyNm', values='count').fillna(0)                 # 방법 1
    temp=temp.pivot_table(index='crdCoBzkNm', columns='sndCtgyNm', values='count', aggfunc='sum')  # 방법 2
    temp['res']=temp.apply('count',axis=1)
    temp=temp.sort_values(by='res',ascending=False)
    temp.to_excel('_res/'+'(python)pivot_check_'+names+'.xlsx')
    return temp[temp['res']>1]    

In [133]:
# 레벨링
def EncodingLabel(x):
    le = LabelEncoder()
    le.fit(x)
    label = le.transform(x)
    return label

---

### 0차 클렌징 처리 - Raw Data 내 문제 해결

#### Problem 1) Y값이 2개인 경우  

|레이블|n수|중복Y값수|처리|근거|
|---|---|---|---|---|
|농축수산품|782|	2|'생활편의시설(유통)' 387건 --->  '음료/식품', 395건으로 통합 처리|농축수산품은 생활편의시설이 아님 + '농산물'이 해당 분류로 구분됨|
|사진관|749|2|'취미생활' 362건 ---> '용역서비스' 387건으로 통합 처리|'사진관.현상소'가 해당 분류로 구분됨|
|선물의집|688|2|'생활편의시설(유통)' 388건 ---> '선물용품' 300건으로 통합 처리|'선물용품점'이 해당 분류로 구분됨|
|수족관|720|2|'오락/문화/관람' 369건 ---> '가구/인테리어/건축자재' 351건으로 통합 처리|'조경/화원/수족관'이 해당 분류로 구분됨|
|안경,콘텍트렌즈|	761|2|'가전제품' 372건 --> '약국및의료기기구매' 389건에 통합 처리|안경은 가전제품이 아님|
|재활용품|442|2|'약국및의료기기구매' 151건 ---> '생활용품' 291건으로 통합 처리|'가정생활용품'이 해당 분류로 구분됨|
|커피전문점|868|2|'음료/식품' 403건 ---> '가족식사' 465건으로 통합 처리|'커피/음료전문점'이 해당 분류로 구분됨|

In [134]:
pivot_check(df,'RawData') # 변경 이전

sndCtgyNm   가구/인테리어/건축자재   가전제품   가족식사  건강식품  결혼/장례서비스지원  공예/토산품  관광(외국인전용점)  \
crdCoBzkNm                                                                     
농축수산품                NaN    NaN    NaN   NaN         NaN     NaN         NaN   
재활용품                 NaN    NaN    NaN   NaN         NaN     NaN         NaN   
안경,콘텍트렌즈             NaN  372.0    NaN   NaN         NaN     NaN         NaN   
온누리상품권               NaN    NaN    NaN   NaN         NaN     NaN         NaN   
수족관                351.0    NaN    NaN   NaN         NaN     NaN         NaN   
커피전문점                NaN    NaN  465.0   NaN         NaN     NaN         NaN   
사진관                  NaN    NaN    NaN   NaN         NaN     NaN         NaN   
선물의집                 NaN    NaN    NaN   NaN         NaN     NaN         NaN   

sndCtgyNm   기념품  기부금  레저/스포츠  ...  의류/잡화  자녀보육  전자상거래  제조/도매업관리  차량서비스/차량연료  \
crdCoBzkNm                    ...                                             
농축수산품       NaN  NaN     NaN  ...    NaN   NaN    NaN       NaN         NaN   
재활용품        NaN  NaN     NaN  ...    NaN   NaN    NaN       NaN         NaN   
안경,콘텍트렌즈    NaN  NaN     NaN  ...    NaN   NaN    NaN       NaN         NaN   
온누리상품권      NaN  NaN     NaN  ...    NaN   NaN    NaN       NaN         NaN   
수족관         NaN  NaN     NaN  ...    NaN   NaN    NaN       NaN         NaN   
커피전문점       NaN  NaN     NaN  ...    NaN   NaN    NaN       NaN         NaN   
사진관         NaN  NaN     NaN  ...    NaN   NaN    NaN       NaN         NaN   
선물의집        NaN  NaN     NaN  ...    NaN   NaN    NaN       NaN         NaN   

sndCtgyNm    취미생활  컴퓨터/통신  학습자재/도서구매  학원수강  res  
crdCoBzkNm                                       
농축수산품         NaN     NaN        NaN   NaN    2  
재활용품          NaN     NaN        NaN   NaN    2  
안경,콘텍트렌즈      NaN     NaN        NaN   NaN    2  
온누리상품권        NaN     NaN        NaN   NaN    2  
수족관           NaN     NaN        NaN   NaN    2  
커피전문점         NaN     NaN        NaN   NaN    2  
사진관         362.0     NaN        NaN   NaN    2  
선물의집          NaN     NaN        NaN   NaN    2  

[8 rows x 37 columns]

In [135]:
df.loc[df['crdCoBzkNm']=='농축수산품','sndCtgyNm']='음료/식품'
df.loc[df['crdCoBzkNm']=='사진관','sndCtgyNm']='용역서비스'
df.loc[df['crdCoBzkNm']=='선물의집','sndCtgyNm']='선물용품'
df.loc[df['crdCoBzkNm']=='수족관','sndCtgyNm']='가구/인테리어/건축자재'
df.loc[df['crdCoBzkNm']=='안경,콘텍트렌즈','sndCtgyNm']='약국및의료기기구매'
df.loc[df['crdCoBzkNm']=='재활용품','sndCtgyNm']='생활용품'
df.loc[df['crdCoBzkNm']=='커피전문점','sndCtgyNm']='가족식사'

In [136]:
pivot_check(df,'RawData_fixed') # 처리 완료!

sndCtgyNm   가구/인테리어/건축자재  가전제품  가족식사  건강식품  결혼/장례서비스지원  공예/토산품  관광(외국인전용점)  \
crdCoBzkNm                                                                   
온누리상품권               NaN   NaN   NaN   NaN         NaN     NaN         NaN   

sndCtgyNm   기념품  기부금  레저/스포츠  ...  의류/잡화  자녀보육  전자상거래  제조/도매업관리  차량서비스/차량연료  \
crdCoBzkNm                    ...                                             
온누리상품권      NaN  NaN     NaN  ...    NaN   NaN    NaN       NaN         NaN   

sndCtgyNm   취미생활  컴퓨터/통신  학습자재/도서구매  학원수강  res  
crdCoBzkNm                                      
온누리상품권       NaN     NaN        NaN   NaN    2  

[1 rows x 37 columns]

---

### 1차 클렌징 처리 - 정규표현식 사용
삭제처리 : 특수문자 일괄 삭제처리

In [137]:
res=[]
for i in range(len(df['crdCoBzkNm'])):
    # 특수문자 제거
    r=re.sub('[-=+,#/\?:^$.@*\"※~&%ㆍ!』\\‘|\(\)\[\]\<\>`\'…》 ／_　．，・內（）]','', df['crdCoBzkNm'][i])
    res.append(r)
df['crdCoBzkNm']=res

---

### 2차 클렌징 처리 - reshape 함수 사용
교정처리 : 명백한 오타인 경우에 한해 교정 처리 적용

In [138]:
df['crdCoBzkNm']=df['crdCoBzkNm'].str.replace('철재가구','철제가구')       # 명백한 오타
df['crdCoBzkNm']=df['crdCoBzkNm'].str.replace('부페','뷔페')               # 명백한 오타
df['crdCoBzkNm']=df['crdCoBzkNm'].str.replace('일식회집','일식횟집')       # 명백한 오타
df['crdCoBzkNm']=df['crdCoBzkNm'].str.replace('차량견인서비스','차량견인') # 차량견인,1건
df['crdCoBzkNm']=df['crdCoBzkNm'].str.replace('캐쥬얼의류','캐주얼의류')   # 명백한 오타
df['crdCoBzkNm']=df['crdCoBzkNm'].str.replace('택시회사','택시')           # 택시회사,1건
df['crdCoBzkNm']=df['crdCoBzkNm'].str.replace('기타광학제품','기타광학품') # 기타광학제품,1건
df['crdCoBzkNm']=df['crdCoBzkNm'].str.replace('렌터카','렌트카')           # 명백한 오타
df['crdCoBzkNm']=df['crdCoBzkNm'].str.replace('보훈매점','보훈매장')       # 보훈매점,1건
df['crdCoBzkNm']=df['crdCoBzkNm'].str.replace('ＣＡＴＶ','CATV')           # 일반적인 영문자가 아님
df['crdCoBzkNm']=df['crdCoBzkNm'].str.replace('ＰＧ','PG')                 # 일반적인 영문자가 아님
df['crdCoBzkNm']=df['crdCoBzkNm'].str.replace('와이셔츠','Y셔츠')          # 용어 통일

In [139]:
pivot_check(df,'regex+reshape_fixed')

sndCtgyNm   가구/인테리어/건축자재  가전제품  가족식사  건강식품  결혼/장례서비스지원  공예/토산품  관광(외국인전용점)  \
crdCoBzkNm                                                                   
문화취미기타               NaN   NaN   NaN   NaN         NaN     NaN         NaN   
민예공예토산품              NaN   NaN   NaN   NaN         NaN   506.0         NaN   
중고품판매점             359.0   NaN   NaN   NaN         NaN     NaN         NaN   
기념품점                 NaN   NaN   NaN   NaN         NaN     NaN         NaN   
수족관                720.0   NaN   NaN   NaN         NaN     NaN         NaN   
온누리상품권               NaN   NaN   NaN   NaN         NaN     NaN         NaN   

sndCtgyNm     기념품  기부금  레저/스포츠  ...  의류/잡화  자녀보육  전자상거래  제조/도매업관리  차량서비스/차량연료  \
crdCoBzkNm                      ...                                             
문화취미기타        NaN  NaN     NaN  ...    NaN   NaN    NaN       NaN         NaN   
민예공예토산품       NaN  NaN     NaN  ...    NaN   NaN    NaN       NaN         NaN   
중고품판매점        NaN  NaN     NaN  ...    NaN   NaN   16.0       NaN         NaN   
기념품점        363.0  NaN     NaN  ...    NaN   NaN    NaN       NaN         NaN   
수족관           NaN  NaN     NaN  ...    NaN   NaN    NaN       NaN         NaN   
온누리상품권        NaN  NaN     NaN  ...    NaN   NaN    NaN       NaN         NaN   

sndCtgyNm    취미생활  컴퓨터/통신  학습자재/도서구매  학원수강  res  
crdCoBzkNm                                       
문화취미기타      510.0     NaN        NaN   NaN    2  
민예공예토산품     361.0     NaN        NaN   NaN    2  
중고품판매점        NaN     NaN        NaN   NaN    2  
기념품점        357.0     NaN        NaN   NaN    2  
수족관           NaN     NaN        NaN   NaN    2  
온누리상품권        NaN     NaN        NaN   NaN    2  

[6 rows x 37 columns]

#### Problem 2) X값이 사실상 동일인데 합쳐보니 Y값이 2개인 경우  
|레이블|n수|중복Y값수|현재상황|근거|처리구상방안|
|---|---|---|---|---|---|
|문화취미기타|861|2|'오락/문화/관람' 351건 vs. '취미생활' 510건|'기타복지문화'가 '오락/문화/관람' 분류로 구분됨|구분이 불가능하여 삭제 요망|
|민예공예토산품|867|2|'공예/토산품' 506건 vs. '취미생활' 361건|'미술공예', '민공예품', '민예공예품'이 '취미생활' 분류로 구분됨 vs. '관광민예선물용품'이 '공예/토산품' 분류로 구분됨|구분이 불가능하여 삭제 요망|
|중고품판매점|375|2|'가구/인테리어/건축자재' 359건 vs. '전자상거래' 16건|'중고물품점', '중고물품판매점'이 '전자상거래' 분류로 구분됨|온라인 매매가 섞인 것으로 보이며 구분 불가능하므로 삭제 요망|
|기념품점|720|2|'기념품' 363건 vs. '취미생활' 357건|'관광기념품점', '수집품기념품'이 '기념품' 분류로 구분됨|'기념품'으로 통합하는 방안 고려|
|수족관|1013|2|'가구/인테리어/건축자재' 720건 vs. '오락/문화/관람' 293건|'조경화원수족관'이 '가구/인테리어/건축자재' 분류로 구분됨|어항과 아쿠아리움 같은 시설이 동일명으로 섞인 것으로 생각되며, 구분 불가능하므로 삭제 요망|

In [140]:
df=df.loc[df['crdCoBzkNm']!='문화취미기타',:]
df=df.loc[df['crdCoBzkNm']!='민예공예토산품',:]
df=df.loc[df['crdCoBzkNm']!='중고품판매점',:]
df.loc[df['crdCoBzkNm']=='기념품점','sndCtgyNm']='기념품'
df=df.loc[df['crdCoBzkNm']!='수족관',:]

In [141]:
pivot_check(df,'regex+reshape+overlap_fixed')

sndCtgyNm   가구/인테리어/건축자재  가전제품  가족식사  건강식품  결혼/장례서비스지원  공예/토산품  관광(외국인전용점)  \
crdCoBzkNm                                                                   
온누리상품권               NaN   NaN   NaN   NaN         NaN     NaN         NaN   

sndCtgyNm   기념품  기부금  레저/스포츠  ...  의류/잡화  자녀보육  전자상거래  제조/도매업관리  차량서비스/차량연료  \
crdCoBzkNm                    ...                                             
온누리상품권      NaN  NaN     NaN  ...    NaN   NaN    NaN       NaN         NaN   

sndCtgyNm   취미생활  컴퓨터/통신  학습자재/도서구매  학원수강  res  
crdCoBzkNm                                      
온누리상품권       NaN     NaN        NaN   NaN    2  

[1 rows x 37 columns]

In [142]:
df

crdCoBzkNm   sndCtgyNm sndCtgy sxClNm  useYymm fstCtgyNm agePrd  no  \
0              양식        가족식사     A01      남   201801      가정친화    10대   1   
1         일반음식점기타        가족식사     A01      남   201801      가정친화    10대   2   
2          일식생선회집        가족식사     A01      남   201801      가정친화    10대   3   
3              중식        가족식사     A01      남   201801      가정친화    10대   4   
4         커피음료전문점        가족식사     A01      남   201801      가정친화    10대   5   
...           ...         ...     ...    ...      ...       ...    ...  ..   
296897   일반치과한방병원       병의원진료     B01      여   201912      건강관리    80대   4   
296898       종합병원       병의원진료     B01      여   201912      건강관리    80대   5   
296899         약국   약국및의료기기구매     B02      여   201912      건강관리    80대   6   
296900       화장품점          미용     D02      여   201912      여가생활    80대   7   
296901       슈퍼마켓  생활편의시설(유통)     D12      여   201912      여가생활    80대   8   

        crdUseAmt  crdUseCnt fstCtgy  
0           12000          1       A  
1          100000          4       A  
2           12000          1       A  
3           30000          2       A  
4           30900          4       A  
...           ...        ...     ...  
296897      49700          1       B  
296898      95100          1       B  
296899     122300          1       B  
296900      17000          1       D  
296901     134360          8       D  

[293786 rows x 11 columns]

In [143]:
# DF 중간 저장
df=df.reset_index(drop=True)
df.to_csv('_res/Dic_process_before.csv',header=True,index=False)

---

### 사전 추가를 통한 단어 인식능력 향상 처리
* 실제 POS 처리 결과 사전 추가 작업이 불필요한 것으로 판단되어 본 과정 패스함

In [144]:
# DF 중간 저장본 로드
# df=pd.read_csv('_res/Dic_process_before.csv',header=0)

---

### NLP 처리 진행
* 이후 정제된 전체 데이터로 진행 예정

In [145]:
# NLP 필요 컬럼만 별도 추출
df_v=df['crdCoBzkNm'].values

In [146]:
df_v

array(['양식', '일반음식점기타', '일식생선회집', ..., '약국', '화장품점', '슈퍼마켓'], dtype=object)

In [147]:
mecab.pos(df_v[0])

[('양식', 'NNG')]

In [149]:
# 토큰화
def getPPS(x):
    res=[]
    pps=['NNG','NNP','NNB','NNBC','NR','NP','VV','VA','VX',
         'VCP','VCN','MM','MAG','MAJ','IC','JKS','JKC','JKG',
         'JKO','JKB','JKV','JKQ','JX','JC','EP','EF','EC',
         'ETN','ETM','XPN','XSN','XSV','XSA','XR','SF',
         'SE','SSO','SSC','SC','SY','SL','SH','SN']  # 모든 품사 지정
    pos=mecab.pos(x)
    #print(pos)
    for po in pos:
        if (po[1] in pps):
            res.append(po[0])
    return res

In [150]:
# for i in range(len(df_v)):
#     r=getPPS(df_v[i])
#     print(r)
#     print('-'*30)

In [151]:
papers=[]
for s in df_v:
    ps=getPPS(s)
    sen=' '.join(ps)
    papers.append(sen)
papers

['양식',
 '일반 음식점 기타',
 '일식 생선 회집',
 '중식',
 '커피 음료 전문점',
 '커피 전문점',
 '패스트푸드',
 '패스트푸드 점',
 '한식',
 '제과점 아이스크림 점',
 '가전제품 점',
 '컴퓨터 소프트웨어',
 '주유소',
 '일반 치과 의원',
 '약국',
 '생명 보험',
 '문방구점',
 '서점',
 'RF 교통 기관 수수료 선취',
 '철도',
 '비디오 및 도서대 여점',
 '공연장 극장',
 '영화관',
 '공공 기관 직영 점',
 '슈퍼마켓',
 '편의점',
 '하나 로 마트',
 '인터넷 쇼핑몰',
 '전자 상거래 안심 클릭 미 적용',
 '전자 상거래 PG',
 '양식',
 '일반 음식점 기타',
 '일식 생선 회집',
 '커피 전문점',
 '패밀리 레스토랑',
 '패스트푸드 점',
 '한식',
 '휴게 음식점',
 '제과점 아이스크림 점',
 '기성 복점',
 '농협 주유소',
 '주유소',
 '약국',
 '자동차 학원',
 '서점',
 '종합 스포츠 센터',
 '화장품 점',
 'RF 교통 기관 수수료 선취',
 '고속버스',
 '철도',
 '공공 기관 직영 점',
 '대형 마트',
 '백화점',
 '슈퍼마켓',
 '편의점',
 '인터넷 쇼핑몰',
 '전자 상거래 안심 클릭 미 적용',
 '전자 상거래 PG',
 '가정용품 수리',
 '기타 휴게 음식점',
 '양식',
 '일반 음식점 기타',
 '일반 음식점 기타',
 '일식 생선 회집',
 '일식 생선 회집',
 '중식',
 '커피 음료 전문점',
 '커피 전문점',
 '패밀리 레스토랑',
 '패스트푸드 점',
 '한식',
 '휴게 음식점',
 '제과점',
 '제과점 아이스크림 점',
 '기성 복점',
 '남여 기성복',
 '악세사리 점',
 '농협 주유소',
 '주유소',
 '차량 정비 부품 인테리어',
 '일반 치과 의원',
 '생명 보험',
 '자동차 학원',
 '문구 용품',
 '문방구점',
 '서점',
 '당구장',
 '볼링장',
 '스포츠 용품

In [76]:
tokenizer=Tokenizer()
tokenizer.fit_on_texts(papers)                     # 단어 인덱스 생성
tm=tokenizer.texts_to_matrix(papers, mode='count') # 카운트기반 매트릭스 생성
tm[:3]

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 1., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [77]:
# 원핫인코딩 처리결과 DF로 저장
df_onehot=pd.DataFrame(tm).loc[:,1:]

In [78]:
# 컬럼에 사용한 Key값 조회
#tokenizer.word_index.keys()

In [79]:
# # Key값을 컬럼명으로 갖는 DF 생성
df_named=pd.DataFrame(tm).loc[:,1:]
df_named.columns = list(tokenizer.word_index.keys())
df_named.astype(int).tail()
df_named.to_csv('_res/df_named.csv',header=True,index=False)

---

### 기존 데이터프레임 추가 컬럼 생성 및 일부 컬럼 제거

In [80]:
# 추가 변수 생성
df['Year']=df['useYymm'].astype('str').str[:4]                         # 연도
df['Mouth']=df['useYymm'].astype('str').str[4:]                        # 월
df['crdUseAmt_D_Cnt']=(df['crdUseAmt']/df['crdUseCnt']).astype('int')  # 카드사용금액/사용횟수

In [81]:
# 기존 변수 삭제 : no, fstCtgy, sndCtgy, useYymm 컬럼
df=df.drop(columns=['no','fstCtgy','sndCtgy','useYymm'],axis=1,inplace=False)

---

### 기본 데이터프레임 컬럼 레벨링 처리

|Code|sndCtgyNm|agePrd|fstCtgyNm|sxClNm|
|---|---|---|---||
|0|'가구/인테리어/건축자재'|'10대'|'가정친화'|'남'|
|1|'가전제품'|'20대'|'건강관리'|'여'|
|2|'가족식사'|'30대'|'여가생활'||
|3|'건강식품'|'40대'|'온누리(지역사랑)상품권'||
|4|'결혼/장례서비스지원'|'50대'|'온누리상품권'||
|5|'공예/토산품'|'60대'|'자기계발'||
|6|'관광(외국인전용점)'|'70대'|||
|7|'기념품'|'80대'|||
|8|'기부금'||||
|9|'레저/스포츠'||||
|10|'미용'||||
|11|'병의원진료'||||
|12|'보험/금융'||||
|13|'생활용품'||||
|14|'생활편의시설(유통)'||||
|15|'선물용품'||||
|16|'숙박시설'||||
|17|'약국및의료기기구매'||||
|18|'여가활용'||||
|19|'여행/교통'||||
|20|'오락/문화/관람'||||
|21|'온누리상품권'||||
|22|'용역서비스'||||
|23|'운송/임대'||||
|24|'유아용품'||||
|25|'음료/식품'||||
|26|'의류/잡화'||||
|27|'자녀보육'||||
|28|'전자상거래'||||
|29|'제조/도매업관리'||||
|30|'차량서비스/차량연료'||||
|31|'취미생활'||||
|32|'컴퓨터/통신'||||
|33|'학습자재/도서구매'||||
|34|'학원수강'||||

In [82]:
# 라벨링 이름 추출 함수
def EncodingLabelName(x):
    le = LabelEncoder()
    le.fit(x)
    label = le.transform(x)
    print(le.classes_)
    return label

In [83]:
# 라벨링 이름 추출 처리
print(set(EncodingLabelName(df['sndCtgyNm'])))
print('-'*100)
print(set(EncodingLabelName(df['agePrd'])))
print('-'*100)
print(set(EncodingLabelName(df['fstCtgyNm'])))
print('-'*100)
print(set(EncodingLabelName(df['sxClNm'])))

['가구/인테리어/건축자재' '가전제품' '가족식사' '건강식품' '결혼/장례서비스지원' '공예/토산품' '관광(외국인전용점)'
 '기념품' '기부금' '레저/스포츠' '미용' '병의원진료' '보험/금융' '생활용품' '생활편의시설(유통)' '선물용품'
 '숙박시설' '약국및의료기기구매' '여가활용' '여행/교통' '오락/문화/관람' '온누리(지역사랑)(개인구매)'
 '온누리상품권(개인구매)' '용역서비스' '운송/임대' '유아용품' '음료/식품' '의류/잡화' '자녀보육' '전자상거래'
 '제조/도매업관리' '차량서비스/차량연료' '취미생활' '컴퓨터/통신' '학습자재/도서구매' '학원수강']
{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35}
----------------------------------------------------------------------------------------------------
['10대' '20대' '30대' '40대' '50대' '60대' '70대' '80대']
{0, 1, 2, 3, 4, 5, 6, 7}
----------------------------------------------------------------------------------------------------
['가정친화' '건강관리' '여가생활' '온누리(지역사랑)상품권' '온누리상품권' '자기계발']
{0, 1, 2, 3, 4, 5}
----------------------------------------------------------------------------------------------------
['남' '여']
{0, 1}


In [84]:
# 라벨링 결과 반영

In [85]:
df[['sndCtgyNm','agePrd','fstCtgyNm','sxClNm']]=df[['sndCtgyNm','agePrd','fstCtgyNm','sxClNm']].apply(EncodingLabel)

In [86]:
# 데이터 프레임 결합

In [87]:
df_final=pd.concat([df,df_onehot],axis=1)
df_final.to_csv('_res/df_final.csv',header=True,index=False)